Refactor Aqara Roller Driver E1 as v2 quirk for more entities

jeverley opened the pull request:
## Proposed change
This PR refactors the existing Aqara Roller Driver E1 quirk into the new v2 structure to expose the configuration and status attributes, below is a summary of changes:

 - Refactor into v2 quirk format
   - Refactor clusters to use AttributeDefs format
 - Expose attributes as entities:
   - **Motor speed** (select)
   - **Charging status** (binary sensor)
   - **Calibration status** (binary sensor)
 - Represent the device as a **_Rollershade_** WindowCoveringType rather than **_Drapery_**
 - Device cluster attribute handling:
   - Reads/reports of `AnalogOutput: present_value` (current position) are replicated into `WindowCovering: current_position_lift_percentage`
   - Writes to `AnalogOutput: present_value` (go to position) do not update the `WindowCovering: current_position_lift_percentage`
   - Reads of `WindowCovering: current_position_lift_percentage` are redirected to `AnalogOutput: present_value`
 - Window Covering command handling:
   - `up_open`, `down_close`, `stop` are handled by `MultistateOutput: present_value`
   - `go_to_lift_percentage` value is written to `AnalogOutput: present_value`
   - The `current_position_lift_percentage` is read prior to returning command response, this ensures the correct state is shown following a change in direction


**Functional enhancements**

- Users can change the Speed using the select entity
- Users can see if the device is Charging
- Users can see if the device is Calibrated
- The cover uses the Rollershade type
- The redundant number entity for current position is disabled by default
- The redundant switch entity is disabled by default

<img width="202" alt="image" src="https://github.com/user-attachments/assets/04eac041-4319-449c-a43c-4a9014c8e4ac" />

## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->

Depends on the following PRs for binary_sensor value conversion:
 - https://github.com/zigpy/zigpy/pull/1540
 - https://github.com/zigpy/zha/pull/360

Depends on the following PR for disabling the redundant entities:
 - https://github.com/zigpy/zigpy/pull/1535
 - https://github.com/zigpy/zha/pull/387

Associated PRs:
 - Fixes cover handling of Rollershade device class (for the correct icon) https://github.com/zigpy/zha/pull/376

Whilst I've validated that is works with my own E1 roller shade devices, I'd like to ensure other users do not encounter any regressions with this refactor.
@dmulcahey / @schwickster / @javicalle / @badrpc / @TheJulianJES I can see in the past you've previously contributed for this device, if you still use it would you mind checking if this updated quirk works as expected without introducing regressions?

## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works

jeverley committed:
Refactor as v2 quirk

- Refactor into v2 quirk
 - Refactor clusters to use AttributeDefs format
 - Represent the device as a Rollershade rather than Drapery
 - Refresh current position after a stop command is issued
 - Expose attributes as entities:
	 - Motor speed (select)
	 - Charging status (binary sensor)
	 - Calibration status (binary sensor)

- Reads to current_position_lift_percentage are redirected to AnalogOutput present_value
 - Writes to AnalogOutput present_value from go_to_lift_percentage commands do not prematurely update the current_position_lift_percentage

jeverley commented:
Have identified an issue with the cover state not reliably changing to open/closed. Investigating this currently.

jeverley commented:
Edit: https://github.com/zigpy/zha/pull/376 now raised to address the below.



@TheJulianJES, @dmulcahey looking at the code in https://github.com/zigpy/zha/blob/dev/zha/application/platforms/cover/__init__.py
it looks to me like the `_determine_state` function and stored target positions might be the cause of the issue, I'm seeing similar issues reported for other devices which would appear to have the same root cause as below:

- https://github.com/home-assistant/core/issues/93836
- https://github.com/home-assistant/core/issues/98933
- https://github.com/zigpy/zha-device-handlers/issues/1397

There was a PR against the old core ZHA repo to try and address this but that looks abandoned https://github.com/home-assistant/core/pull/105467

I've tested the following scenarios:

**1. _go_to_lift_percentage_ has never been called, _down_close_ is called, then the device reports _position_or_tilt_ of 0**
_target_lift_position == None
position_or_tilt == 0
returns **STATE_CLOSED**

**2. _go_to_lift_percentage_ has never been called, _up_open_ is called, then the device reports _position_or_tilt_ of 100**
_target_lift_position == None
position_or_tilt == 100
returns: **STATE_OPEN**

**3. _go_to_lift_percentage_ is called (go to 0), then the device reports _position_or_tilt_ of 0**
_target_lift_position == 0
position_or_tilt == 0
returns **STATE_CLOSED**

**4. _go_to_lift_percentage_ is called (go to 100), then the device reports _position_or_tilt_ of 100**
_target_lift_position == 100
position_or_tilt == 100
returns **STATE_OPEN**

**5. _go_to_lift_percentage_ is called (go to 50), then the device reports _position_or_tilt_ of 50**
_target_lift_position == 50
position_or_tilt == 50
returns **STATE_OPEN**

**6. _go_to_lift_percentage_ has previously been called (stored value is 50), _down_close_ is called, then the device reports _position_or_tilt_ of 0**
_target_lift_position == 50
position_or_tilt == 0
returns **STATE_CLOSED**

**7. _go_to_lift_percentage_ has previously been called (stored value is 50), _up_open_ is called, then the device reports _position_or_tilt_ of 100**
_target_lift_position == 50
position_or_tilt == 100
returns **None, state remains as STATE_OPENING**

**8. _go_to_lift_percentage_ has previously been called (stored value is 50) and current position is 100, _stop_ is called**
_target_lift_position == 100 (the function async_stop_cover sets it to the current cover position)
position_or_tilt == 100
returns **STATE_OPEN**

I'd suggest we can handle this in one of 3 ways:

1. set the stored _target_lift_position to 100 or 0 when calling the _up_open_ or _down_close_ commands respectively
2. clear the stored _target_lift_position when calling the _up_open_ or _down_close_ commands
3. modify the _determine_state function to return STATE_OPEN when `self.is_opening == True and position_or_tilt == 100` is true


Below is the current function for reference:
```
    def _determine_state(self, position_or_tilt, is_lift_update=True) -> None:
        """Determine the state of the cover.

        In HA None is unknown, 0 is closed, 100 is fully open.
        In ZCL 0 is fully open, 100 is fully closed.
        Keep in mind the values have already been flipped to match HA
        in the WindowCovering cluster handler
        """
        if is_lift_update:
            target = self._target_lift_position
            current = self.current_cover_position
        else:
            target = self._target_tilt_position
            current = self.current_cover_tilt_position

        if position_or_tilt == 0:
            self._state = (
                STATE_CLOSED
                if is_lift_update
                else STATE_OPEN
                if self.current_cover_position is not None
                and self.current_cover_position > 0
                else STATE_CLOSED
            )
            return
        if target is not None and target != current:
            # we are mid transition and shouldn't update the state
            return
        self._state = STATE_OPEN
```

jeverley commented:
I've been working on changes to the ZHA cover entity hander to address the underlying status issue outlined above (it also affects the existing v1 quirk code for this and other devices).

I'll raise a bug and PR to address these over the weekend, current WIP code here:
https://github.com/jeverley/zha/blob/cover-platform-entity-status/zha/application/platforms/cover/__init__.py

jeverley commented:
@TheJulianJES if I'm reading https://github.com/zigpy/zigpy/pull/1540 I believe I should be able to use the proposed `attribute_converter` functionality to remap the Aqara values for the charging binary_sensor?

If so I'll remove the custom converter I've implemented in this quirk once your PR is merged.

jeverley commented:
@TheJulianJES the `AnalogOutput` cluster `present_value` on this device is used to write the intended target position, whereas reads/reports on the attribute report the device current position.

Right now ZHA exposes this as a separate number entity in addition to the cover entity.
In the UI the number entity is redundant since interactions & current position can all be handled through the cover.

When adding custom entities in a v2 quirk we're able to mark them as disabled by default, do you think it would be beneficial to extend this to regular cluster attributes?
This would allow us to present devices more uniformly, without exposing redundant entities arising from manufacturers' non-standard implementations.

Critically this would be different to entirely removing the cluster/marking an attribute as unsupported in the cluster, since we'd still want to remap attribute updates/writes from the cover commands to the cluster.

TheJulianJES commented:
> When adding custom entities in a v2 quirk we're able to mark them as disabled by default, do you think it would be beneficial to extend this to regular cluster attributes?

With https://github.com/zigpy/zigpy/pull/1535, that should be allowed in the future. Do note that the ZHA part wasn't PR'd yet, so it won't do anything for now.
`attribute_converter` support also isn't merged, yet.

jeverley committed:
Hide redundant entities by default

The AnalogOutput cluster is used to write target position to the device.
The present_value reported/read provides the current position/starting position when the device is in motion.

Both of these functions are re-mapped into the ZCL WindowCovering cluster, so the number entity created is redundant.

This also pre-emptively hides the MultistateOutput cluster, though there is currently no cluster handler implemented for this.

Apply pre-commit auto fixes

jeverley committed:
Replace bool conversion function with attribute_converter

Depends on the following PRs:
 - https://github.com/zigpy/zigpy/pull/1540
 - https://github.com/zigpy/zha/pull/360

jeverley committed:
Remove non-applicable blocked status

The `MultistateOutput: present value` only ever reports the last written command (even if the device has since stopped/changed direction).

jeverley committed:
Negative test scenarios for WriteAwareUpdateAttribute class

jeverley committed:
Handle unsupported commands cleanly

jeverley commented:
@TheJulianJES I think this should be ready for merging, I'll raise a PR for the zha cover status issue outlined in https://github.com/zigpy/zha-device-handlers/pull/3686#issuecomment-2578907234 separately (it's not a pre-req for this change).

puddly commented on zhaquirks/xiaomi/aqara/roller_curtain_e1.py:
```diff
@@ -99,218 +245,104 @@ async def command(
+    .prevent_default_entity_creation(
+        endpoint_id=1, cluster_id=MultistateOutput.cluster_id
+    )
+    .removes(OnOff.cluster_id)
```
For this use case, why remove the cluster entirely? I'm hoping to replace uses of `removes` with `prevent_default_entity_creation`.

jeverley replied:
As far as I can tell this cluster is non-functional.

Interacting with the switch results in an error, so there would be no benefit exposing it.
If it's an optional entity a user might be tempted to enable it and raise a bug report that the switch doesn't work.

puddly replied:
I think `removes(OnOff.cluster_id)` can be replaced with `prevent_default_entity_creation(cluster_id=OnOff.cluster_id)`. Removing the cluster prevent prevents automatic entity discovery, which is more explicit with `prevent_default_entity_creation` and would allow you to still interact with the `OnOff` cluster via cluster commands if necessary.

jeverley replied:
I'd usually agree, but the device itself is responding with the below error to **any** command to OnOff:

Failed to perform the action zha/issue_zigbee_cluster_command. Failed to issue cluster command with status: <Status.UNSUPPORTED_CLUSTER: 195>

jeverley replied:
@puddly if we're planning to deprecate `removes` then happy to change this to use `prevent_default_entity_creation`.

jeverley replied:
I've updated the PR to use `prevent_default_entity_creation` https://github.com/zigpy/zha-device-handlers/pull/3686/commits/872442a3be297f93968ec0d9519d8cb2b33dc43f

jeverley replied:
Hi @puddly, did you have any other feedback on these changes?

puddly replied:
Not at the moment, sorry. The above method depends on https://github.com/zigpy/zha/pull/365 to be finished and there were a few hiccups getting it merged.

jeverley committed:
Improve docstring annotations and use t.uint8_t type for current_position_lift_percentage updates

jeverley commented:
I've now raised https://github.com/zigpy/zha/pull/376 to address the ZHA logic issues I outlined in https://github.com/zigpy/zha-device-handlers/pull/3686#issuecomment-2578907234.
This isn't a pre-requisite for merging this PR.

jeverley committed:
Use prevent_default_entity_creation instead of removes for OnOff

Per @puddly 's recommendation in https://github.com/zigpy/zha-device-handlers/pull/3686#discussion_r1943100229

jeverley committed:
Read current position after issuing commands

The current_position_lift_percentage now is read prior to returning command responses to ensure that ZHA has the correct position during changes in direction/stopping.

Apply pre-commit auto fixes and fix tests.

jeverley commented:
Will update the tests following https://github.com/zigpy/zha-device-handlers/pull/3686/commits/9e5b67ac9a7749c1bfb3f95827462da4cede735c later this evening.

Edit: _fixed and ready with https://github.com/zigpy/zha-device-handlers/pull/3686/commits/d342b64b61c25395e5ae7bb76f9f94ddd3a6bc62_

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

TheJulianJES commented on zhaquirks/xiaomi/aqara/roller_curtain_e1.py:
```diff
@@ -94,223 +240,121 @@ async def command(
+        fallback_name="Charging",
+        attribute_converter=lambda x: True
+        if x == AqaraRollerDriverCharging.true
+        else False,
```
```suggestion
        attribute_converter=lambda x: x == AqaraRollerDriverCharging.true,
```

TheJulianJES commented on zhaquirks/xiaomi/aqara/roller_curtain_e1.py:
```diff
@@ -2,64 +2,194 @@
+class AqaraRollerDriverCharging(t.enum8):
+    """Aqara roller driver charging status attribute values."""
+
+    true = 0x01
+    false = 0x02
```
This should really be (upper-case) `True` and `False`, but that would conflict.
Maybe something like `Charging` and `Unplugged`? (assuming "not charging" => "unplugged").

Technically, we don't even need this at all, as we could just parse `x == 1` for the charging binary sensor and have no enum.

TheJulianJES reviewed (approved):
Not sure if I'm a huge fan of the `is_write`/`WriteAwareUpdateAttribute` stuff and `RedirectAttributes` (we do need a good way of doing this in general), but it seems to look fine. Since the ZHA PR is in, we should probably get this in as well.

I'm assuming that you have tested it quite a bit. Unfortunately, I don't have any cover devices, so unable to test anything. Again, looks fine, so I'll quickly correct the two small things I noted and get it in for the beta tomorrow. If there are any huge issues, we can always fix them during the beta or revert if necessary.

TheJulianJES committed:
Rename `AqaraRollerDriverCharging` values, improve `attribute_converter`

TheJulianJES committed:
Merge branch 'dev' into lumi.curtain.acn002-v2-quirk
